include/oneapi/dpl/pstl/utils.h

```diff
@@ -818,30 +822,109 @@ __shars_upper_bound(_Acc __acc, _Size __first, _Size __last, const _Value& __val
+struct __iterator_value_type
+{
+    using __type = void;
+};
+
+template <typename _Iterator>
+struct __iterator_value_type<_Iterator, std::void_t<typename std::iterator_traits<std::decay_t<_Iterator>>::value_type>>
+{
+    using __type = typename std::iterator_traits<std::decay_t<_Iterator>>::value_type;
+};
+
 template <typename _Iterator1, typename _Iterator2, typename = void>
-struct __is_equality_comparable : std::false_type
+struct __has_same_value_types : std::false_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2>
+struct __has_same_value_types<_Iterator1, _Iterator2,
+                              std::enable_if_t<std::is_same_v<typename __iterator_value_type<_Iterator1>::__type,
+                                                              typename __iterator_value_type<_Iterator2>::__type>>>
+    : std::true_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct __has_equality_op : std::false_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2>
+struct __has_equality_op<_Iterator1, _Iterator2,
+                         std::void_t<decltype(std::declval<_Iterator1>() == std::declval<_Iterator2>())>>
+    : std::true_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct __is_equality_self_comparable : std::false_type
 {
 };
 
-// All with implemented operator ==
 template <typename _Iterator1, typename _Iterator2>
-struct __is_equality_comparable<
-    _Iterator1, _Iterator2,
-    std::void_t<decltype(::std::declval<::std::decay_t<_Iterator1>>() == ::std::declval<::std::decay_t<_Iterator2>>())>>
+struct __is_equality_self_comparable<_Iterator1, _Iterator2,
+                                     std::enable_if_t<std::conjunction_v<__has_same_value_types<_Iterator1, _Iterator2>,
+                                                                         __has_equality_op<_Iterator1, _Iterator2>>>>
     : std::true_type
 {
 };
 
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct __is_equality_comparable : std::false_type
+{
+};
+
+template <typename _Iterator1, typename _Iterator2>
+struct __is_equality_comparable<_Iterator1, _Iterator2,
+                                std::enable_if_t<__is_equality_self_comparable<_Iterator1, _Iterator2>::value>>
+    : std::conditional_t<std::conjunction_v<__base_iterator_type<_Iterator1>, __base_iterator_type<_Iterator2>>,
+                         __is_equality_self_comparable<_Iterator1, _Iterator2>,
+                         __is_equality_comparable<typename __base_iterator_type<_Iterator1>::__type,
+                                                  typename __base_iterator_type<_Iterator2>::__type>>
+{
+};
+#endif // _ONEDPL_CPP20_CONCEPTS_PRESENT
+
+template <typename _Iterator1, typename _Iterator2>
+inline constexpr bool __is_equality_comparable_v =
+    __is_equality_comparable_impl::__is_equality_comparable<_Iterator1, _Iterator2>::value;
```

Comment:
Its a little strange to me to have this within the `__is_equality_compareable_impl` namespace.  
Id recommend moving this final `bool` outside to the `oneapi::dpl::__internal` namespace.

Reply:
The idea was that it is one place used thing,
Fixed: moved outside of namespace `__is_equality_comparable_impl`